core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -434,18 +440,46 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     }.toMap
   }
 
-  def removeExpiredOffsets(startMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
-    val expiredOffsets = offsets
-      .filter {
+  def removeExpiredOffsets(currentTimestamp: Long, offsetRetentionMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
+
+    def getExpiredOffsets(baseTimestamp: CommitRecordMetadataAndOffset => Long): Map[TopicPartition, CommitRecordMetadataAndOffset] = {
+      offsets.filter {
         case (topicPartition, commitRecordMetadataAndOffset) =>
-          commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp < startMs && !pendingOffsetCommits.contains(topicPartition)
-      }
-      .map {
-        case (topicPartition, commitRecordOffsetAndMetadata) =>
-          (topicPartition, commitRecordOffsetAndMetadata.offsetAndMetadata)
-      }
+          !pendingOffsetCommits.contains(topicPartition) && {
+            commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp match {
+              case None =>
+                // current version with no per partition retention
+                currentTimestamp - baseTimestamp(commitRecordMetadataAndOffset) >= offsetRetentionMs
+              case Some(expireTimestamp) =>
+                // older versions with explicit expire_timestamp field => old expiration semantics is used
+                currentTimestamp >= expireTimestamp
+            }
+          }
+      }.toMap
+    }
+
+    val expiredOffsets: Map[TopicPartition, CommitRecordMetadataAndOffset] = protocolType match {
+      case Some(protocol) if is(Empty) =>
+        // no consumer exists in the group =>
+        // if retention period has passed since group became Empty, expire all offsets with no pending offset commit
+        getExpiredOffsets(commitRecordMetadataAndOffset => {currentStateTimestamp.get})
```

Comment:
nit: you can simplify 
```scala
        getExpiredOffsets(_ => currentStateTimestamp.get)
```
Also, are you sure the `get` is safe? What happens for a group that is loaded from an old version of the metadata schema since it has no current timestamp? It seems like we would load `None` in that case.